ms2deepscore/SpectrumBinner.py

```diff
@@ -69,8 +71,14 @@ def from_json(cls, json_str: str):
                               binner_dict["mz_max"], binner_dict["mz_min"],
                               binner_dict["peak_scaling"],
                               binner_dict["allowed_missing_percentage"])
+        # loads in all the classes in MetadataFeatureGenerator.py
         if "additional_metadata" in binner_dict:
-            spectrum_binner.additional_metadata = binner_dict["additional_metadata"]
+            possible_metadata_classes = import_module("ms2deepscore.MetadataFeatureGenerator")
```

Comment:
This makes me think.
As it is here users can only use classes defined in ms2deepscore. But one of the main advantages of using a base class `MetadataFeatureGenerator` is that everyone can build their own metadata-to-feature conversions.
Passing custom objects won't work either, because that will be difficult to convert to json and save.
Another problem with this approach is: If people store their model (which includes a binner), but we later decide to change some rules, e.g. m/z divided by 5000 instead of 1000, then loading a model will not fail nor complain but won't work as expected.

One alternative I can think of is to work with more generic conversion mechanisms, say: feature-to-onehot, feature-to-binary, feature-scaler and store all needed factors and choices in a dictionary internally. We could build a class for each such a conversion. The feature-to-binary would need to contain attributes such as `metadata_field`, `entries_becoming_zero`, `entries_becoming_one`...

Reply:
e.g.
```python
class StandardScaler:
    def __init__(self, metadata_field: str, mean: float, std: float = None):
        self.metadata_field = metadata_field
        self.mean = mean
        self.std = std

    def generate_features(self, metadata):
        feature = metadata.get(self.metadata_field, None)
        assert self.metadata_field is not None, f"Metadata entry for {self.metadata_field} is missing."
        if self.std:
             return (feature - self.mean) / std
        return feature - self.meann
```

Reply:
We should then add a simple `to_json` method to the respective base class that will export what is needed to reconstruct. I guess that would only be the attributes plus the class name.

Reply:
I think that makes a lot of sense. In the end this would not need that many different classes and it makes it robust and easy for users to try other things without the need to clone MS2Deepscore and adding classes to the code. I will have a look into this.